### PR TITLE
JCLOUDS-951: Remove inheritance in ProfitBrick's AutoValue classes

### PR DESCRIPTION
**Pull request by devcsrj**
Continuing from the ProfitBricks adapter progress left 3 months ago, here's a PR for [JCLOUDS-951](https://issues.apache.org/jira/browse/JCLOUDS-951).

- Inheritance-based domain classes removed
- Now using `@AutoValue.Builder` (ausuum)
- Added missed validations on some domain classes

### CODE REVIEW

**Comment by zack-shoylev** [comment]
Great cleanup, but might take a bit to review.

File: profitbricks/src/main/java/org/jclouds/profitbricks/domain/DataCenter.java [hunk 1]
```diff
@@ -49,81 +46,55 @@
...
-//   @Nullable public abstract List<LoadBalancer> loadBalancers();
-
-   public static DataCenter create(String id, String name, int version, ProvisioningState state, Location location, List<Server> servers,
-           List<Storage> storages) {
-      return new AutoValue_DataCenter(id, name, version, state, location,
-              servers != null ? ImmutableList.copyOf(servers) : Lists.<Server>newArrayList(),
-              storages != null ? ImmutableList.copyOf(storages) : Lists.<Storage>newArrayList());
-   }
 
+//   @Nullable public abstract List<LoadBalancer> loadBalancers();
```
Comment (zack-shoylev) [inline profitbricks/src/main/java/org/jclouds/profitbricks/domain/DataCenter.java]: ?

File: profitbricks/src/main/java/org/jclouds/profitbricks/domain/DataCenter.java [hunk 2]
```diff
@@ -49,81 +46,55 @@
...
-         this.version = version;
-         return this;
-      }
+      public abstract Builder location(Location location);
 
-      public Builder servers(List<Server> servers) {
-         this.servers = servers;
-         return this;
-      }
+      public abstract Builder servers(List<Server> servers);
```
Comment (zack-shoylev) [inline profitbricks/src/main/java/org/jclouds/profitbricks/domain/DataCenter.java]: List vs ImmutableList?

same below.

Comment (zack-shoylev) [inline profitbricks/src/main/java/org/jclouds/profitbricks/domain/DataCenter.java]: To be more precise, should probably use autoBuild and copyOf, like other places in this PR?

File: profitbricks/src/main/java/org/jclouds/profitbricks/domain/Firewall.java [hunk 3]
```diff
@@ -59,228 +56,77 @@ public static Protocol fromValue(String value) {
...
-               this.rules.add(rule);
-               return this;
-            }
-
-            public AddRulePayload build() {
-               return AddRulePayload.create(nicId, rules);
-            }
-         }
-      }
+   public Builder toBuilder() {
```
Comment (zack-shoylev) [inline profitbricks/src/main/java/org/jclouds/profitbricks/domain/Firewall.java]: What about abstract Builder toBuilder(); and have autovalue take care of it?

**Comment by zack-shoylev** [comment]
Some small nits, but looks good.

File: profitbricks/src/main/java/org/jclouds/profitbricks/domain/DataCenter.java (hunk 2 above)
Comment (devcsrj) [inline profitbricks/src/main/java/org/jclouds/profitbricks/domain/DataCenter.java]: Like [this](https://github.com/devcsrj/jclouds-labs/commit/a4cced5ecadc56d6b02de5e18b9e6249d90761d4)?

**Comment by devcsrj** [comment]
Thanks for the review @zack-shoylev ! I have addressed the issues, and will squash the commits on your signal

**Comment by zack-shoylev** [comment]
Looks good to merge! Let's squash the commits.

**Comment by devcsrj** [comment]
Rebased and squashed.

**Comment by zack-shoylev** [comment]
Hopefully just a hickup and I can merge after; there were some issues running the tests.

**Comment by zack-shoylev** [comment]
/scratch/jenkins/workspace/jclouds-labs-pull-requests/profitbricks/src/test/java/org/jclouds/profitbricks/features/LoadbalancerApiLiveTest.java:61: error: cannot find symbol
              .loadBalancerName("testName")
              ^
  symbol:   method loadBalancerName(String)
  location: class Builder
/scratch/jenkins/workspace/jclouds-labs-pull-requests/profitbricks/src/test/java/org/jclouds/profitbricks/features/LoadbalancerApiLiveTest.java:92: error: cannot find symbol
      LoadBalancer.Request.RegisterPayload payload = LoadBalancer.Request.registerBuilder()
                                                                         ^
  symbol:   method registerBuilder()
  location: class Request
/scratch/jenkins/workspace/jclouds-labs-pull-requests/profitbricks/src/test/java/org/jclouds/profitbricks/features/LoadbalancerApiLiveTest.java:107: error: cannot find symbol
      LoadBalancer.Request.DeregisterPayload payload = LoadBalancer.Request.deregisterBuilder()
                                                                           ^
  symbol:   method deregisterBuilder()
  location: class Request
/scratch/jenkins/workspace/jclouds-labs-pull-requests/profitbricks/src/test/java/org/jclouds/profitbricks/features/LoadbalancerApiLiveTest.java:121: error: cannot find symbol
              .loadBalancerName("whatever")
              ^
  symbol:   method loadBalancerName(String)
  location: class Builder
4 errors
10 warnings

"loadBalancerName" was renamed to "name", etc.

Just a few small fixes.

**Comment by devcsrj** [comment]
Ooh. Missed that. Should've built it one more time on my machine. Anyway, here ya go!

**Comment by zack-shoylev** [comment]
merged

**Pull request closed without merging**